components/cdc/src/endpoint.rs

```diff
@@ -413,6 +419,52 @@ impl<T: 'static + RaftStoreRouter<RocksEngine>> Endpoint<T> {
         }
     }
 
+    fn get_current_resolved_ts(
+        &self,
+        region_id: u64,
+        observer_id: ObserveID,
+    ) -> impl Future<Output = Option<TimeStamp>> {
+        let pd_client = self.pd_client.clone();
+        let cm = self.concurrency_manager.clone();
+        let raft_router = self.raft_router.clone();
+        let scheduler = self.scheduler.clone();
+        let regions = vec![(region_id, observer_id)];
+
+        async move {
+            let mut min_ts = pd_client.get_tso().await.unwrap_or_default();
+
+            cm.update_max_ts(min_ts); // TODO do we need this here?
+            if let Some(min_mem_lock_ts) = cm.global_min_lock_ts() {
+                if min_mem_lock_ts < min_ts {
+                    min_ts = min_mem_lock_ts;
+                }
+            }
+
+            let valid_regions =
+                Self::region_resolved_ts_raft(regions, &scheduler, raft_router, min_ts).await;
+            if valid_regions.len() == 0 {
+                return None;
+            }
+
+            // send a barrier to the scheduler
+            let (tx, rx) = tokio::sync::oneshot::channel();
+            if let Err(e) = scheduler.schedule(Task::Barrier(Box::new(move || {
+                // TODO any risk with unwrap?
+                tx.send(()).unwrap();
+            }))) {
+                warn!("cdc send barrier failed"; "err" => ?e, "min_ts" => min_ts);
+                return None;
+            }
+
+            if let Err(e) = rx.await {
+                warn!("cdc barrier error"; "err" => ?e, "min_ts" => min_ts);
+                return None;
+            }
```

Comment:
Why do we need a barrier?

Reply:
The whole incremental scan process now by-passes the scheduler, so we have to make sure all real-time events that have been produced by this point have been processed before proceeding, or otherwise the resolved ts might not be correct.